### PR TITLE
LUN-126: Recursive diciplines

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

File: src/components/ExerciseHeader.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import React, { useEffect, useState } from 'react'
-import Modal from './Modal'
+import ConfirmationModal from './Modal'
```
Comment (klinzo) [inline src/components/ExerciseHeader.tsx:2]: Should we rename the file as well then?

File: src/constants/endpoints.ts [hunk 2]
```diff
@@ -1,19 +1,12 @@
 import { Article } from './data'
 
-export interface ProfessionType {
+export interface DisciplineType {
```
Comment (klinzo) [inline src/constants/endpoints.ts:3]: Do we have a glossary somewhere for these type of terms? I am really confused all the time what is a document, word, discipline, profession and so on. This is not an app specific topic though, but a project wide. Should I mention this in ~lunes-general?

File: src/hooks/useLoadDisciplines.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  return { ...serverResponse, data: formattedServerResponse }
+}
+
+export const useLoadDisciplines = (parent: DisciplineType | null): ReturnType<DisciplineType[]> => {
+  let url: string
+  if (parent === null) {
+    url = ENDPOINTS.professions.all
+  } else {
+    url = `${!parent.isLeaf ? ENDPOINTS.professions.all : ENDPOINTS.subCategories.all}/${parent?.id}`
+  }
```
Comment (klinzo) [inline src/hooks/useLoadDisciplines.ts:37]: To make it clearer you could extract the different urls to separate constants.
```suggestion
  const url = parent === null ? ENDPOINTS.professions.all : `${!parent.isLeaf ? ENDPOINTS.professions.all : ENDPOINTS.subCategories.all}/${parent?.id}`
```

File: src/hooks/useLoadDisciplines.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  const disciplines: ReturnType<ServerResponse[]> = useLoadFromEndpoint(url)
```
Comment (klinzo) [inline src/hooks/useLoadDisciplines.ts:38]:
```suggestion
  const disciplines = useLoadFromEndpoint<ServerResponse[]>(url)
```

File: src/hooks/useLoadDisciplines.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    serverResponse.data?.map(item => ({
+      ...item,
+      numberOfChildren:
+        item.total_discipline_children > 0
+          ? item.total_discipline_children
+          : item.total_training_sets > 0
+          ? item.total_training_sets
+          : item.total_documents > 0
+          ? item.total_documents
+          : 0,
```
Comment (klinzo) [inline src/hooks/useLoadDisciplines.ts:25]: Isn't this the same as:
```suggestion
        item.total_discipline_children ||  item.total_training_sets ||  item.total_documents
```

File: src/constants/endpoints.ts [hunk 6]
```diff
@@ -41,10 +34,10 @@ export type DocumentsType = DocumentType[]
 
 export const ENDPOINTS = {
   professions: {
-    all: '/disciplines'
+    all: '/disciplines_by_level'
   },
   subCategories: {
-    all: '/training_set/:id'
+    all: '/training_set'
```
Comment (klinzo) [inline src/constants/endpoints.ts:40]: Why is there no `id` anymore here?

I am really confused with the naming here as well, probably nothing for this PR though. I think we should be consistent. +1 for a glossary and consistent naming :D

File: src/hooks/useLoadDisciplines.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      isLeaf: item.total_discipline_children <= 0
```
Comment (klinzo) [inline src/hooks/useLoadDisciplines.ts:26]: Why `<=`?
```suggestion
      isLeaf: item.total_discipline_children === 0
```

File: src/navigation/Navigator.tsx [hunk 8]
```diff
@@ -49,8 +49,9 @@ const Stack = createStackNavigator<RoutesParamsType>()
 
 const Navigator = (): JSX.Element => {
   const [isPressed, setIsPressed] = React.useState(false)
+  const [isHomeButtonPressed, setIsHomeButtonPressed] = React.useState(false)
```
Comment (klinzo) [inline src/navigation/Navigator.tsx:52]:
```suggestion
  const [isPressed, setIsPressed] = React.useState<boolean>(false)
  const [isHomeButtonPressed, setIsHomeButtonPressed] = React.useState<boolean>(false)
```

File: src/navigation/Navigator.tsx [hunk 9]
```diff
@@ -74,36 +86,51 @@ const Navigator = (): JSX.Element => {
...
+              !!route.params.extraParams.parentTitle
+            )
           }
           name='ProfessionSubcategory'
           component={ProfessionSubcategoryScreen}
         />
         <Stack.Screen
           options={({ route, navigation }: any) =>
-            defaultOptions(route.params.extraParams.disciplineTitle, BackButton, navigation, 'ProfessionSubcategory')
+            defaultOptions(route.params.extraParams.disciplineTitle, BackButton, navigation, true)
```
Comment (klinzo) [inline src/navigation/Navigator.tsx:104]: Why did you remove the `ProfessionSubcategory` string? Not needed anymore?

File: src/routes/vocabulary-list/VocabularyListScreen.tsx [hunk 10]
```diff
@@ -46,21 +46,6 @@ interface VocabularyListScreenPropsType {
...
-        <TouchableOpacity
-          onPress={() => navigation.popToTop()}
-          onPressIn={() => setIsHomeButtonPressed(true)}
-          onPressOut={() => setIsHomeButtonPressed(false)}
-          activeOpacity={1}>
-          {isHomeButtonPressed ? <HomeButtonPressed /> : <Home />}
-        </TouchableOpacity>
-      )
-    })
-  }, [navigation, isHomeButtonPressed])
```
Comment (klinzo) [inline src/routes/vocabulary-list/VocabularyListScreen.tsx:63]: Nice refactoring :heart_eyes:

File: src/routes/ProfessionScreen.tsx [hunk 11]
```diff
@@ -104,19 +91,19 @@ const ProfessionScreen = ({ navigation }: ProfessionScreenPropsType): ReactEleme
...
+  const handleNavigation = (item: DisciplineType): void => {
     setSelectedId(item.id)
     navigation.navigate('ProfessionSubcategory', {
       extraParams: {
-        disciplineID: item.id,
-        disciplineTitle: item.title,
-        disciplineIcon: item.icon
+        module: {
+          ...item
+        }
```
Comment (klinzo) [inline src/routes/ProfessionScreen.tsx:106]:
```suggestion
        module: item
```

File: src/routes/ProfessionSubcategoryScreen.tsx [hunk 12]
```diff
@@ -77,51 +77,36 @@ interface ProfessionSubcategoryScreenPropsType {
...
     const selected = item.id === selectedId
     const descriptionStyle = selected ? styles.clickedItemDescription : styles.description
     const badgeStyle = selected ? styles.clickedItemBadgeLabel : styles.badgeLabel
+    const description = module.isLeaf
+      ? item.numberOfChildren === 1
+        ? labels.home.word
+        : labels.home.words
+      : item.numberOfChildren === 1
+      ? labels.home.unit
+      : labels.home.units
```
Comment (klinzo) [inline src/routes/ProfessionSubcategoryScreen.tsx:109]: Indentation

**Review by klinzo — CHANGES_REQUESTED** [review]

File: src/constants/endpoints.ts (hunk 2 above)
Comment (SteffiStuffel) [reply src/constants/endpoints.ts:3]: Yes you are right, it is really confusing and we were talking about this already in lunes calls, but nothing already.
Currently the api has the endpoints:
- disciplines -> trainings_sets -> documents (wich have words)
In the app we have:
- profession -> subprofessions -> documents
Non-technical used naming:
- Bereichen -> Module

As with this PR in the frontend we merge disiplines and trainingssets into one type. I created a ticket and suggested the name "module" for it, but not really happy with it tough, so feed free to suggest an other name for disciplines/trainingssets (actually these are just like "folders" where the documents are inside)
https://issues.tuerantuer.org/browse/LUN-142

But feel also free to open a discussion in ~lunes-general

File: src/hooks/useLoadDisciplines.ts (hunk 5 above)
Comment (SteffiStuffel) [reply src/hooks/useLoadDisciplines.ts:25]: yes it is, and so much nicer :+1:

File: src/constants/endpoints.ts (hunk 6 above)
Comment (SteffiStuffel) [reply src/constants/endpoints.ts:40]: In useLoadDisicplines we append the id, but just appending it is easier, than searching for `:id` and replace it, isn't it?

File: src/navigation/Navigator.tsx (hunk 9 above)
Comment (SteffiStuffel) [reply src/navigation/Navigator.tsx:104]: Yes, not needed.

File: src/routes/ProfessionSubcategoryScreen.tsx (hunk 12 above)
Comment (SteffiStuffel) [reply src/routes/ProfessionSubcategoryScreen.tsx:109]: its prettier :unamused: 
made it a one-liner

Comment (SteffiStuffel) [reply src/routes/ProfessionSubcategoryScreen.tsx:109]: oke is also not working...

File: src/constants/endpoints.ts (hunk 2 above)
Comment (klinzo) [reply src/constants/endpoints.ts:3]: Okay, thanks for explaining :+1: And `subCategories` in `endpoints.ts` is just a legacy of the old structure and basically the same as a training set, i.e. a module with documents as children? Will open a discussion :)

Comment (klinzo) [reply src/constants/endpoints.ts:3]: Maybe just only using the term profession/disclipine would also work instead of renaming to module. That would probably make the purpose a little bit clearer.

Comment (SteffiStuffel) [reply src/constants/endpoints.ts:3]: oke, i will adjust the ticket so it will be renamed to discipline.

**Comment by klinzo** [comment]
Where do we have recursive disciplines already, i.e. where can this be tested?

File: src/constants/endpoints.ts (hunk 2 above)
Comment (klinzo) [reply src/constants/endpoints.ts:3]: I like that, that way we have nearly the same wording as the api as well :)

**Review by klinzo — APPROVED** [review]
Seems to work fine :tada: One thing I noticed: If I navigate to a 1st level discipline, e.g. `Metall, Elektro, Maschinen` the home button in the header is not displayed, is this because the behavior is the same as for the back button? Maybe I'd show it anyway as it may be confusing for users. But not really sure about that.

**Comment by SteffiStuffel** [comment]
> Seems to work fine  One thing I noticed: If I navigate to a 1st level discipline, e.g. `Metall, Elektro, Maschinen` the home button in the header is not displayed, is this because the behavior is the same as for the back button? Maybe I'd show it anyway as it may be confusing for users. But not really sure about that.

This was a design decision, so would just keep it.

**Comment by SteffiStuffel** [comment]
> Where do we have recursive disciplines already, i.e. where can this be tested?
> Edit: Seems to be testable at `Sonstige Bereiche` > `Test Sub 1`.
> Weird thing I noticed: It shows `28 Modules`, but there actually only two (or three if you count recursively). @SteffiStuffel probably an cms problem right?

Yeah. Lukes fixed this already, but wasn't deployed on test-cms, when you tested. I will check out if it now works fine.
